crates/bevy_ecs/src/storage/resource_non_send.rs

```diff
@@ -0,0 +1,522 @@
+
+            result
+        }
+
+        fn run_indirect<F, T>(&mut self, f: F) -> T
+        where
+            F: FnOnce(&mut ThreadLocals) -> T + Send,
+            T: Send + 'static,
+        {
+            let ThreadLocalAccess::Indirect(ref mut sender) = self.access else { unreachable!() };
+
+            let (result_tx, result_rx) = std::sync::mpsc::sync_channel(1);
+            let f = |tls: &mut ThreadLocals| {
+                tls.update_change_tick();
+                let saved = std::mem::replace(&mut tls.last_tick, *self.last_run);
+                // we want to propagate to caller instead of panicking in the main thread
+                let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| f(tls)));
+                tls.last_tick = saved;
+
+                *self.last_run = tls.last_tick;
+                result_tx.send(result).unwrap();
+            };
+
+            let f: Box<dyn FnOnce(&mut ThreadLocals) + Send> = Box::new(f);
+
+            // SAFETY: `scope` blocks the calling thread until `f` completes,
+            // so any borrows within `f` will remain valid until then.
+            let f: Box<dyn FnOnce(&mut ThreadLocals) + Send + 'static> =
+                unsafe { std::mem::transmute(f) };
+
+            // Send task to the main thread.
+            sender
+                .send_task(f)
+                .unwrap_or_else(|_| panic!("receiver missing"));
+
+            // Wait to receive result back from the main thread.
+            result_rx.recv().unwrap().unwrap()
+        }
+    }
+
+    // SAFETY: This system param does not borrow any data from the world.
+    unsafe impl SystemParam for ThreadLocal<'_, '_> {
+        type State = ThreadLocalState;
+        type Item<'w, 's> = ThreadLocal<'w, 's>;
+
+        fn init_state(
+            world: &mut crate::prelude::World,
+            system_meta: &mut crate::system::SystemMeta,
+        ) -> Self::State {
+            let component_id =
+                crate::system::ResMut::<ThreadLocalAccessor>::init_state(world, system_meta);
+
+            ThreadLocalState {
+                component_id,
+                last_run: Tick::new(0),
+            }
+        }
+
+        unsafe fn get_param<'world, 'state>(
+            state: &'state mut Self::State,
+            system_meta: &crate::system::SystemMeta,
+            world: UnsafeWorldCell<'world>,
+            curr_tick: Tick,
+        ) -> Self::Item<'world, 'state> {
+            let mut accessor = crate::system::ResMut::<ThreadLocalAccessor>::get_param(
+                &mut state.component_id,
+                system_meta,
+                world,
+                curr_tick,
+            );
+
+            let access = if std::thread::current().id() == accessor.owning_thread {
+                // SAFETY: We are on the owning thread and pointer is valid.
+                let ptr = accessor.direct;
+                unsafe { ThreadLocalAccess::Direct(&mut *ptr) }
```

Comment:
This is unsound since `ThreadLocalAccess` wasn't created with a mutable reference.

Reply:
For now, I'll change it to an `unsafe` method that requires passing in a `*mut ThreadLocals` directly.